### PR TITLE
chore(metrics): Separate out iterating over single and multi-value tags

### PR DESCRIPTION
**Pull request by bruceg**
The `MetricTags` implementation has a single implementation of iteration over its values that previously only had single-valued results, but now has multi-valued results. These new semantics are unsuitable for most current uses of metric tags, so separate these out into iterating over tags with a single value and interating over all values of all tags.

Epic: #14742

### CODE REVIEW

**Review by neuronull — COMMENTED** [review]
Just a question on what appears to be changed behavior but otherwise, without too much context into the metric tags code, not seeing any issues stand out.

File: lib/vector-core/src/event/metric/series.rs [hunk 1]
```diff
@@ -131,8 +131,11 @@ impl fmt::Display for MetricSeries {
         write_word(fmt, &self.name.name)?;
         write!(fmt, "{{")?;
         if let Some(tags) = &self.tags {
-            write_list(fmt, ",", tags.iter(), |fmt, (tag, value)| {
-                write_word(fmt, tag).and_then(|()| write!(fmt, "={:?}", value))
+            write_list(fmt, ",", tags.iter_all(), |fmt, (tag, value)| {
+                write_word(fmt, tag).and_then(|()| match value {
+                    Some(value) => write!(fmt, "={:?}", value),
+                    None => Ok(()),
```
[the remark this answers is not in the document]
Comment (bruceg) [reply lib/vector-core/src/event/metric/series.rs:137]: A value of `None` represents a "bare" tag. In the above, the tag name is already written out, so we don't need to write out a tag value (as opposed to an empty tag, which will still be represented). So, this should preduce `bare_tag,empty_tag="",something="value"`.

File: src/sinks/datadog/metrics/encoder.rs [hunk 2]
```diff
@@ -393,8 +393,11 @@ fn get_namespaced_name(metric: &Metric, default_namespace: &Option<Arc<str>>) ->
 
 fn encode_tags(tags: &MetricTags) -> Vec<String> {
     let mut pairs: Vec<_> = tags
-        .iter()
-        .map(|(name, value)| format!("{}:{}", name, value))
+        .iter_all()
+        .map(|(name, value)| match value {
+            Some(value) => format!("{}:{}", name, value),
+            None => name.into(),
```
[the remark this answers is not in the document]
Comment (bruceg) [reply src/sinks/datadog/metrics/encoder.rs:399]: Datadog ingest represents bare tags without a separator, which this is intended to produce. We can't actually handle bare tags _quite_ yet (#15177) but that is the plan.

File: lib/vector-core/src/event/metric/series.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (neuronull) [reply lib/vector-core/src/event/metric/series.rs:137]: :+1: Thanks for the additional context.

**Review by neuronull — APPROVED** [review]
LGTM

**Pull request merged**